libs/agno/agno/document/reader/csv_reader.py

```diff
@@ -50,6 +53,91 @@ def read(self, file: Union[Path, IO[Any]], delimiter: str = ",", quotechar: str
+        """
+        try:
+            if isinstance(file, Path):
+                if not file.exists():
+                    raise FileNotFoundError(f"Could not find file: {file}")
+                logger.info(f"Reading async: {file}")
+                async with aiofiles.open(file, mode="r", encoding="utf-8", newline="") as file_content:
+                    content = await file_content.read()
+                    file_content_io = io.StringIO(content)
+            else:
+                logger.info(f"Reading uploaded file async: {file.name}")
+                file.seek(0)
+                file_content_io = io.StringIO(file.read().decode("utf-8"))  # type: ignore
+
+            csv_name = Path(file.name).stem if isinstance(file, Path) else file.name.split(".")[0]
+
+            # First pass to count rows and detect if the file is small enough for single-document mode
+            file_content_io.seek(0)
+            csv_reader = csv.reader(file_content_io, delimiter=delimiter, quotechar=quotechar)
+            rows = list(csv_reader)
+            total_rows = len(rows)
+
+            # For very small files, just create a single document
+            if total_rows <= 10:  # Adjust threshold as needed
+                csv_content = " ".join(", ".join(row) for row in rows)
+                documents = [
+                    Document(
+                        name=csv_name,
+                        id=f"{csv_name}_1",
+                        content=csv_content,
+                    )
+                ]
+            else:
+                # Divide rows into pages
+                pages = []
+                for i in range(0, total_rows, page_size):
+                    pages.append(rows[i : i + page_size])
+
+                async def _process_page(page_number: int, page_rows: List[List[str]]) -> Document:
+                    """Process a page of rows into a document"""
+                    start_row = (page_number - 1) * page_size + 1
+                    page_content = " ".join(", ".join(row) for row in page_rows)
+
+                    return Document(
+                        name=csv_name,
+                        id=f"{csv_name}_page{page_number}",
+                        meta_data={"page": page_number, "start_row": start_row, "rows": len(page_rows)},
+                        content=page_content,
+                    )
+
+                # Process pages in parallel
+                documents = await asyncio.gather(
+                    *[_process_page(page_number, page) for page_number, page in enumerate(pages, start=1)]
+                )
+
+            # Apply chunking if needed
+            if self.chunk:
+
+                async def _chunk_document(doc: Document) -> List[Document]:
+                    return await asyncio.to_thread(self.chunk_document, doc)
+
+                # Process chunks in parallel
+                chunked_lists = await asyncio.gather(*[_chunk_document(doc) for doc in documents])
```

Comment:
Lets make this change to the base reader. Add a async_chunk_document method the creates the thread (as a workaround for now) and then just call and gather that.
You can update the PDF reader with the same